### ISSUE TITLE
Perf: NonGeneric.HashTable default constructor 4x slower on Linux

### ISSUE THREAD

**Issue by ianhays** [dotnet/corefx#3590]
Perf test:
```
[Benchmark]
public void ctor()
{
    foreach (var iteration in Benchmark.Iterations)
        using (iteration.StartMeasurement())
        {
            for (int i = 0; i < 40000; i++)
            {
                new Hashtable(); new Hashtable(); new Hashtable(); new Hashtable(); new Hashtable();
                new Hashtable(); new Hashtable(); new Hashtable(); new Hashtable(); new Hashtable();
                new Hashtable(); new Hashtable(); new Hashtable(); new Hashtable(); new Hashtable();
                new Hashtable(); new Hashtable(); new Hashtable(); new Hashtable(); new Hashtable();
            }
        }
}
```

Linux results:
```
        <performance runid="System.Collections.Tests.Perf_HashTable.ctor-Linux" etl="/home/ianha/uptonogood/results/System.Collections.Tests.Perf_HashTable.ctor-Linux.csv">
          <metrics>
            <Duration displayName="Duration" unit="msec" />
          </metrics>
          <iterations>
            <iteration index="0" Duration="88.903400003910065" />
            <iteration index="1" Duration="85.48089998960495" />
            <iteration index="2" Duration="86.384900003671646" />
            <iteration index="3" Duration="120.60969999432564" />
            <iteration index="4" Duration="91.006500005722046" />
            <iteration index="5" Duration="90.413500010967255" />
            <iteration index="6" Duration="86.7703999876976" />
            <iteration index="7" Duration="106.35569998621941" />
            <iteration index="8" Duration="106.56749999523163" />
            <iteration index="9" Duration="85.525199979543686" />
            <iteration index="10" Duration="86.409900009632111" />
            <iteration index="11" Duration="86.196299999952316" />
          </iterations>
        </performance>
```

Windows results:
```
        <performance runid="System.Collections.Tests.Perf_HashTable.ctor-WindowsCore" etl="D:\git\corefx\bin\tests\Windows_NT.AnyCPU.Release\System.Collections.NonGeneric.Tests\dnxcore50\System.Collections.Tests.Perf_HashTable.ctor-WindowsCore.etl">
          <metrics>
            <Duration displayName="Duration" unit="msec" />
          </metrics>
          <iterations>
            <iteration index="0" Duration="20.593591045927496" />
            <iteration index="1" Duration="19.145835822048639" />
            <iteration index="2" Duration="18.775771568288405" />
            <iteration index="3" Duration="18.844196391788046" />
            <iteration index="4" Duration="18.386890488065774" />
            <iteration index="5" Duration="18.788031015832075" />
            <iteration index="6" Duration="18.766648258488431" />
            <iteration index="7" Duration="20.691096419414521" />
            <iteration index="8" Duration="20.270283754892034" />
            <iteration index="9" Duration="18.497795722821252" />
            <iteration index="10" Duration="18.67798109137027" />
            <iteration index="11" Duration="18.28168732193501" />
            <iteration index="12" Duration="18.674844953626575" />
            <iteration index="13" Duration="18.681972539407752" />
            <iteration index="14" Duration="18.63664109383933" />
            <iteration index="15" Duration="18.660874885495332" />
            <iteration index="16" Duration="18.321886905741053" />
            <iteration index="17" Duration="18.596156406601949" />
            <iteration index="18" Duration="18.423383727265445" />
            <iteration index="19" Duration="18.845051702081719" />
            <iteration index="20" Duration="18.289670218009974" />
            <iteration index="21" Duration="18.75153777663229" />
            <iteration index="22" Duration="18.791737360438219" />
            <iteration index="23" Duration="18.338137801322318" />
            <iteration index="24" Duration="18.924880662831129" />
            <iteration index="25" Duration="18.726448674682388" />
            <iteration index="26" Duration="18.807418049156922" />
            <iteration index="27" Duration="18.282827735660021" />
            <iteration index="28" Duration="18.899221354018891" />
            <iteration index="29" Duration="18.595301096308276" />
            <iteration index="30" Duration="18.733576260463678" />
            <iteration index="31" Duration="18.739848535951069" />
            <iteration index="32" Duration="18.597296820326847" />
            <iteration index="33" Duration="19.242485885241877" />
            <iteration index="34" Duration="18.782043843775682" />
            <iteration index="35" Duration="19.010696795636932" />
            <iteration index="36" Duration="18.923170042243783" />
            <iteration index="37" Duration="20.555387186140479" />
            <iteration index="38" Duration="19.868858123694508" />
            <iteration index="39" Duration="19.6114097252771" />
            <iteration index="40" Duration="19.75111040658885" />
            <iteration index="41" Duration="19.914189569262817" />
            <iteration index="42" Duration="20.092094110361813" />
            <iteration index="43" Duration="20.034503217249721" />
            <iteration index="44" Duration="23.98489636062618" />
            <iteration index="45" Duration="21.004425090356563" />
            <iteration index="46" Duration="20.392308023466285" />
            <iteration index="47" Duration="19.883968605550535" />
            <iteration index="48" Duration="20.176484726011267" />
            <iteration index="49" Duration="22.660876025909147" />
            <iteration index="50" Duration="21.773349044433189" />
            <iteration index="51" Duration="19.444339114565764" />
            <iteration index="52" Duration="18.541131444371103" />
          </iterations>
        </performance>
```

Probably somehow related to #3588. Confirmed results with Stopwatch.

**Comment by stephentoub** [comment]
For all of these issues about ctor performance with objects that are purely managed code, I don't expect there's going to be anything to fix on the corefx side of thing.  It'd be good to come with some canonical example that can be used to open an issue against coreclr instead.  Hopefully we can fix it on the coreclr side of things.  If not, we might learn about certain operations that have different performance profiles on the two systems, such that we then might need to customize the implementation based on those performance differences, but I don't think it'll come to that.

**Comment by stephentoub** [comment]
I believe this is due to https://github.com/dotnet/coreclr/issues/1708

**Comment by kouvel** [comment]
Partially fixed by dotnet/coreclr#1746. Here are my numbers in iterations per second, where each iteration consists of 10,000 allocations:
- Hashtable (Windows): 3200
- Hashtable (Linux): 829 (baseline), 1465 (fixed)
Leaving the issue open since there is still a considerable gap.

**Comment by kouvel** [comment]
Almost all of the remaining difference is in GetThread. There isn't an easy fix for this, so we are currently not planning on addressing this. We can take another look if we see this to be a significant issue in user scenarios.
